File: Java/src/gov/nasa/larcfm/ACCoRD/CriticalVectors.java
/*
 * CriticalVectors.java 
 * Release: ACCoRDj-2.b (08/22/10) 
 *
 * Contact: Cesar Munoz (cesar.a.munoz@nasa.gov)
 * NASA LaRC
 * http://shemesh.larc.nasa.gov/people/cam/ACCoRD
 * 
 * Copyright (c) 2011-2016 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

package gov.nasa.larcfm.ACCoRD;

import static gov.nasa.larcfm.ACCoRD.Consts.*;
import java.util.ArrayList;
import java.util.List;

import gov.nasa.larcfm.Util.*;

public class CriticalVectors {

  public static List<Vect2> tracks(Vect3 s, Vect3 vo, Vect3 vi,
      double D, double H, double B, double T) {
    Vect2 s2  = s.vect2();
    Vect2 vo2 = vo.vect2();
    Vect2 vi2 = vi.vect2();

    if (Util.almost_equals(vo.z,vi.z) && Math.abs(s.z) < H) {
      return CriticalVectors2D.tracks(s2,vo2,vi2,D,B,T);
    } else if (!Util.almost_equals(vo.z,vi.z)) {
      double b = Math.max(Vertical.Theta_H(s.z,vo.z-vi.z,Entry,H),B);
      double t = Math.min(Vertical.Theta_H(s.z,vo.z-vi.z,Exit,H),T);
      if (b < t) {
        return CriticalVectors2D.tracks(s2,vo2,vi2,D,b,t); 
      }
    }
    return new ArrayList<Vect2>();
  }

  public static List<Vect2> tracks(Vect3 s, Vect3 vo, Vect3 vi,
      double D, double H, double T) {
    return tracks(s,vo,vi,D,H,0,T);
  }

  public static List<Vect2> groundSpeeds(Vect3 s, Vect3 vo, Vect3 vi,
      double D, double H, double B, double T) {
    Vect2 s2  = s.vect2();
    Vect2 vo2 = vo.vect2();
    Vect2 vi2 = vi.vect2();

    if (Util.almost_equals(vo.z,vi.z) && Math.abs(s.z) < H) {
      return CriticalVectors2D.groundSpeeds(s2,vo2,vi2,D,B,T);
    } else if (!Util.almost_equals(vo.z,vi.z)) {
      double b = Math.max(Vertical.Theta_H(s.z,vo.z-vi.z,Entry,H),B);
      double t = Math.min(Vertical.Theta_H(s.z,vo.z-vi.z,Exit,H),T);
      if (b < t) {
        return CriticalVectors2D.groundSpeeds(s2,vo2,vi2,D,b,t); 
      }
    }
    return new ArrayList<Vect2>();
  }

  public static List<Vect2> groundSpeeds(Vect3 s, Vect3 vo, Vect3 vi,
      double D, double H, double T) {
    return groundSpeeds(s,vo,vi,D,H,0,T);
  }

  private static List<Double> verticalSpeeds1D(double sz, double viz,
      double H, double B, double T) {
    List<Double> vss = new ArrayList<Double>();
    Vertical nvoz;
    if (Math.abs(sz) < H && B > 0) {
      for (int eps=-1;eps <= 1; eps+=2) {
        nvoz = Vertical.vs_circle_at(sz,viz,B,eps,Exit,H);
        if (!nvoz.undef()) 
          vss.add(nvoz.z);
      }
    } else if (Math.abs(sz) >= H && B > 0) {
      nvoz = Vertical.vs_circle_at(sz,viz,B,-Util.sign(sz),Exit,H);
      if (!nvoz.undef()) 
        vss.add(nvoz.z);
    }
    if (Math.abs(sz) >= H) {
      nvoz = Vertical.vs_circle_at(sz,viz,T,Util.sign(sz),Entry,H);
      if (!nvoz.undef()) 
        vss.add(nvoz.z);
    }
    return vss;
  }

  public static List<Double> verticalSpeeds(Vect3 s, Vect3 vo, Vect3 vi,
      double D, double H, double B, double T) {
    Vect2 s2  = s.vect2();
    Vect2 vo2 = vo.vect2();
    Vect2 vi2 = vi.vect2();
//f.pln("CriticalVectors.verticalSpeeds "+s + " "+vo+" "+vi);      

    if (vo2.almostEquals(vi2) && s2.sqv() < Util.sq(D)) {
//f.pln("CriticalVectors.verticalSpeeds 1 "+verticalSpeeds1D(s.z,vi.z,H,B,T));      
      return verticalSpeeds1D(s.z,vi.z,H,B,T);
    } else {
      Vect2 v2 = vo2.Sub(vi2);
      if (Horizontal.Delta(s2,v2,D) > 0) {
        double b = Math.max(Horizontal.Theta_D(s2,v2,Entry,D),B);
        double t = Math.min(Horizontal.Theta_D(s2,v2,Exit,D),T);
//f.pln("CriticalVectors.verticalSpeeds 2 "+verticalSpeeds1D(s.z,vi.z,H,b,t)+" b="+b+" t="+t);      
        if (b < t) {
//f.pln("CriticalVectors.verticalSpeeds 3 "+verticalSpeeds1D(s.z,vi.z,H,b,t));      
          return verticalSpeeds1D(s.z,vi.z,H,b,t); 
        }
      }
    }
    return new ArrayList<Double>();  
  }

  public static List<Double> verticalSpeeds(Vect3 s, Vect3 vo, Vect3 vi,
      double D, double H, double T) {
    return verticalSpeeds(s,vo,vi,D,H,0,T);
  }
  
}

File: Modules/ACCoRD/src/CriticalVectors.cpp
/*
 * CriticalVectors.cpp
 * Release: ACCoRD++-2.b (08/22/10)
 *
 * Contact: Cesar Munoz (cesar.a.munoz@nasa.gov)
 * NASA LaRC
 * http://shemesh.larc.nasa.gov/people/cam/ACCoRD
 *
 * Copyright (c) 2011-2019 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

#include "CriticalVectors.h"
#include "Util.h"
#include <cmath>

namespace larcfm {
  
  std::list<Vect2> CriticalVectors::tracks(const Vect3& s, const Vect3& vo, const Vect3& vi,
			  const double D, const double H, const double B, const double T) {
    Vect2 s2  = s.vect2();
    Vect2 vo2 = vo.vect2();
    Vect2 vi2 = vi.vect2();
    
    if (Util::almost_equals(vo.z,vi.z) && std::abs(s.z) < H) {
      return CriticalVectors2D::tracks(s2,vo2,vi2,D,B,T);
    } else if (!Util::almost_equals(vo.z,vi.z)) {
      double b = Util::max(Vertical::Theta_H(s.z,vo.z-vi.z,larcfm::Entry,H),B);
      double t = Util::min(Vertical::Theta_H(s.z,vo.z-vi.z,larcfm::Exit,H),T);
      if (b < t) {
        return CriticalVectors2D::tracks(s2,vo2,vi2,D,b,t);
      }
    }
    std::list<Vect2> trks;
    return trks;
  }
  
  std::list<Vect2> CriticalVectors::tracks(const Vect3& s, const Vect3& vo, const Vect3& vi,
			  const double D, const double H, const double T) {
    return tracks(s,vo,vi,D,H,0,T);
  }

  std::list<Vect2> CriticalVectors::groundSpeeds(const Vect3& s, const Vect3& vo, const Vect3& vi,
				const double D, const double H, const double B, const double T) {
    Vect2 s2  = s.vect2();
    Vect2 vo2 = vo.vect2();
    Vect2 vi2 = vi.vect2();
    if (Util::almost_equals(vo.z,vi.z) && std::abs(s.z) < H) {
      return CriticalVectors2D::groundSpeeds(s2,vo2,vi2,D,B,T);
    } else if (!Util::almost_equals(vo.z,vi.z)) {
      double b = Util::max(Vertical::Theta_H(s.z,vo.z-vi.z,larcfm::Entry,H),B);
      double t = Util::min(Vertical::Theta_H(s.z,vo.z-vi.z,larcfm::Exit,H),T);
      if (b < t) {
        return CriticalVectors2D::groundSpeeds(s2,vo2,vi2,D,b,t);
      }
    }
    std::list<Vect2> gss;
    return gss;
  }

  std::list<Vect2> CriticalVectors::groundSpeeds(const Vect3& s, const Vect3& vo, const Vect3& vi,
				const double D, const double H, const double T) {
    return groundSpeeds(s,vo,vi,D,H,0,T);
  }

  std::list<double> verticalSpeeds1D(const double sz, const double viz,
				     const double H, const double B, const double T) {
    std::list<double> vss;
    Vertical nvoz;
    if (std::abs(sz) < H && B > 0) {
      for (int eps=-1;eps <= 1; eps+=2) {
        nvoz = Vertical::vs_circle_at(sz,viz,B,eps,larcfm::Exit,H);
        if (!nvoz.undef()) 
          vss.push_back(nvoz.z);
      }
    } else if (std::abs(sz) >= H && B > 0) {
      nvoz = Vertical::vs_circle_at(sz,viz,B,-sign(sz),larcfm::Exit,H);
      if (!nvoz.undef()) 
        vss.push_back(nvoz.z);
    }
    if (std::abs(sz) >= H) {
      nvoz = Vertical::vs_circle_at(sz,viz,T,sign(sz),larcfm::Entry,H);
      if (!nvoz.undef()) 
        vss.push_back(nvoz.z);
    }
    return vss;
  }

  std::list<double> CriticalVectors::verticalSpeeds(const Vect3& s, const Vect3& vo, const Vect3& vi,
				   const double D, const double H, const double B, const double T) {
    Vect2 s2  = s.vect2();
    Vect2 vo2 = vo.vect2();
    Vect2 vi2 = vi.vect2();
    
    if (vo2.almostEquals(vi2) && s2.sqv() < sq(D)) {
      return verticalSpeeds1D(s.z,vi.z,H,B,T);
    } else {
      Vect2 v2 = vo2 - vi2;
      if (Horizontal::Delta(s2,v2,D) > 0) {
        double b = Util::max(Horizontal::Theta_D(s2,v2,larcfm::Entry,D),B);
        double t = Util::min(Horizontal::Theta_D(s2,v2,larcfm::Exit,D),T);
        if (b < t) {
          return verticalSpeeds1D(s.z,vi.z,H,b,t); 
        }
      }
    }
    std::list<double> vss;
    return vss;
  }

  std::list<double> CriticalVectors::verticalSpeeds(const Vect3& s, const Vect3& vo, const Vect3& vi,
				   const double D, const double H, const double T) {
    return verticalSpeeds(s,vo,vi,D,H,0,T);
  }
  
}
